packages/wallet/src/redux/indirect-funding/player-a/reducer.ts

```diff
@@ -32,180 +30,191 @@ import {
 import { WalletEvent } from 'magmo-wallet-client';
 import { isfundingAction } from '../../direct-funding/actions';
 import { addHex } from '../../../utils/hex-utils';
+import { ProtocolStateWithSharedData, SharedData } from '../../protocols';
 
 export function playerAReducer(
-  state: walletStates.Initialized,
+  state: ProtocolStateWithSharedData<states.PlayerAState>,
   action: actions.indirectFunding.Action,
-): walletStates.Initialized {
-  if (!state.indirectFunding) {
-    return state;
-  }
-
-  if (state.indirectFunding.player !== PlayerIndex.A) {
-    return state;
-  }
-
-  switch (state.indirectFunding.type) {
+): ProtocolStateWithSharedData<states.PlayerAState> {
+  switch (state.protocolState.type) {
     case states.WAIT_FOR_APPROVAL:
-      return waitForApprovalReducer(state, action);
+      return waitForApprovalReducer(
+        state as ProtocolStateWithSharedData<states.WaitForApproval>,
```

Comment:
This should not require a type assertion. Either the input type to `waitForApprovalReducer` should be relaxed, or we should figure out how to properly do the discriminated union.

Reply:
Something that would make this easier would be to change the signature of these reducers to:
```ts
function waitForApprovalReducer(
    protocolState: WaitForApproval,
    sharedState: SharedState,
    action: IndirectFundingAction
): ReturnVal {
  // ...
}

type ReturnVal = ProtocolStateWithSharedData<InsufficientFundingState>;
```